aspnetcore/blazor/forms-validation.md

```diff
@@ -223,12 +230,413 @@ In the following example:
+
+Business logic validation can be accomplished by using a [custom validation component](#custom-validation-components):
+
+* Process model validation in the component.
+* Pass field validation errors in a dictionary to the custom validation component.
+
+In the following example:
+
+* The custom validation component shown in the [Custom validation components](#custom-validation-components) section is used.
+* The validation requires that the value of the ship description (`Description`) isn't `null` or empty if the user selects the `Defense` ship classification (`Classification`).
+
+When validation messages are set in the component, they're added to the form's <xref:Microsoft.AspNetCore.Components.Forms.ValidationMessageStore> and shown in the <xref:Microsoft.AspNetCore.Components.Forms.EditForm>:
+
+```csharp
+@page "/FormsValidation"
+@using BlazorSample.Shared
+
+<h1>Starfleet Starship Database</h1>
+
+<h2>New Ship Entry Form</h2>
+
+<EditForm Model="@starship" OnValidSubmit="HandleValidSubmit">
+    <DataAnnotationsValidator />
+    <CustomValidator @ref="customValidator" />
+    <ValidationSummary />
+
+    ...
+
+</EditForm>
+
+@code {
+    private CustomValidator customValidator;
+    private Starship starship = new Starship() { ProductionDate = DateTime.UtcNow };
+
+    private void HandleValidSubmit()
+    {
+        customValidator.ClearErrors();
+
+        var errors = new Dictionary<string, List<string>>();
+
+        if (starship.Classification == "Defense" &&
+                string.IsNullOrEmpty(starship.Description))
+        {
+            errors.Add(nameof(starship.Description),
+                new List<string>() { "For a 'Defense' ship classification, " +
+                "'Description' is required." });
+        }
+
+        if (errors.Count() > 0)
+        {
+            customValidator.DisplayErrors(errors);
+        }
+        else
+        {
+            // Process the form
+        }
+    }
+}
+```
+
+## Server validation
+
+Server validation can be accomplished with a custom server validation component:
+
+* Process client-side validation in the form.
+* When the form passes client-side validation (<xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnValidSubmit> is called), send the <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Model?displayProperty=nameWithType> to a backend server API for form processing.
+* Process model validation on the server.
+* If model validation passes on the server, process the form and send back a success status code (*200 - OK*). If model validation fails, return a failure status code (*400 - Bad Request*) and a dictionary of field validation errors.
+* Either disable the form on success or display the errors.
+
+The following example is based on:
+
+* A hosted Blazor solution created by the Blazor Hosted project template.
+* The *Starfleet Starship Database* form example in the preceding [Built-in forms components](#built-in-forms-components) section.
```

Comment:
nit: typo

Reply:
Where? ...... I probably need another cup of ☕ .... I don't see it.

Reply:
nevermind, I had too many of them I think :)